File: reconcile/report.py
"""Everything this run says out loud: annotations, masking, the job summary.

Remote strings reach all of it - a runner's name, an ix failure reason, an
error quoting either - so `clean` is not decoration. A newline opens a fresh
line, which is where `::` workflow commands parse and where a summary table
takes another row.
"""

from __future__ import annotations

import os
import re
from typing import Any


# A newline in a log line opens a fresh line, where `::` workflow commands
# parse and where a summary table takes another row. Plenty of what we print
# is chosen remotely: a runner's name, an ix failure_reason or status, an
# error message quoting either.
CONTROL_CHARACTERS = re.compile(r"[\x00-\x1f\x7f-\x9f]")

def clean(value: Any) -> str:
    """One line, safe to print: a remote string cannot forge output."""
    return CONTROL_CHARACTERS.sub(" ", str(value))


def log_error(message: str) -> None:
    """An Actions error annotation - surfaced on the run, not buried in logs."""
    print(f"::error::{clean(message)}")


def log_warning(message: str) -> None:
    """An Actions warning annotation."""
    print(f"::warning::{clean(message)}")


def write_summary(rows: list[tuple[str, str, str]]) -> None:
    """Append a per-member outcome table to the job summary, when in Actions."""
    path = os.environ.get("GITHUB_STEP_SUMMARY")
    if not path or not rows:
        return
    lines = ["", "| member | action | outcome |", "| --- | --- | --- |"]
    # A cell carries remote text: a pipe would break the table open and a
    # newline would forge a whole row of it.
    lines += [
        "| " + " | ".join(clean(cell).replace("|", "\\|") for cell in row) + " |"
        for row in rows
    ]
    with open(path, "a", encoding="utf-8") as handle:
        handle.write("\n".join(lines) + "\n")

File: src/report.ts
/** Everything this run says out loud: annotations, masking, the job summary.
 *
 * Remote strings reach all of it - a runner's name, an ix failure reason, an
 * error quoting either - so `clean` is not decoration: a newline opens a
 * fresh line, which is where `::` workflow commands parse and where a
 * summary table takes another row. */

const CONTROL_CHARACTERS = /[\x00-\x1f\x7f-\x9f]/g

/** One line, safe to print: a remote string cannot forge output. */
export function clean(value: unknown): string {
  return String(value).replace(CONTROL_CHARACTERS, " ")
}

/** An Actions error annotation - surfaced on the run, not buried in logs. */
export function logError(message: string): void {
  console.log(`::error::${clean(message)}`)
}

export function logWarning(message: string): void {
  console.log(`::warning::${clean(message)}`)
}

/** Redact a credential from everything this process prints from here on,
 * tracebacks included. Print it before the credential can reach any output. */
export function mask(secret: string): void {
  console.log(`::add-mask::${secret}`)
}

/** Append a per-subject outcome table to the job summary, when in Actions. */
export async function writeSummary(
  rows: readonly (readonly [string, string, string])[],
): Promise<void> {
  const path = process.env.GITHUB_STEP_SUMMARY
  if (!path || rows.length === 0) return
  const cell = (value: string) => clean(value).replaceAll("|", "\\|")
  const lines = [
    "",
    "| subject | action | outcome |",
    "| --- | --- | --- |",
    ...rows.map((row) => `| ${row.map(cell).join(" | ")} |`),
    "",
  ]
  const existing = await Bun.file(path).text().catch(() => "")
  await Bun.write(path, existing + lines.join("\n"))
}
